src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php

```diff
@@ -325,7 +355,17 @@ private function validateAndDenormalize(string $currentClass, string $attribute,
 
                 $childContext = $this->createChildContext($context, $attribute);
                 if ($this->serializer->supportsDenormalization($data, $class, $format, $childContext)) {
-                    return $this->serializer->denormalize($data, $class, $format, $childContext);
+                    try {
+                        return $this->serializer->denormalize($data, $class, $format, $childContext, $accumulatingContext);
+                    } catch (NotDenormalizableValueException | InvalidArgumentException $e) {
+                        if (isset($context[self::COLLECT_ALL_ERRORS]) && $context[self::COLLECT_ALL_ERRORS]) {
+                            $accumulatingContext[$attribute][] = $e->getMessage();
```

Comment:
You should probably merge caught exception with the existing one. It's why collecting messages not enough, it would be better to collect exceptions themselves.

The resulting document (to send to the client) should be something like:

```
attribute1: error message
attribute2.innerAttribute1: foo
attribute2.innerAttribute2: bar
```

Reply:
At the moment errors live in a nested array like the following one(using your example):

```
$errors = [
    'attribute1' => [
        0 => 'error message',
    ],
    'attribute2' => [
        'innerAttribute1' => [
            0 => 'foo',
        ],
        'innerAttribute2' => [
            0 => 'bar',
        ]
    ],
];
```

Which should be fairly easy to serialize in the form you suggest.. do you think it'd be better to make it flat from the beginning? i.e.:
```
$errors = [
    'attribute1' => [
        0 => 'error message',
    ],
    'attribute2.innerAttribute1' => [
        0 => 'foo',
    ],
    'attribute2.innerAttribute2' => [
        0 => 'bar',
    ],
];
```